packages/wrangler/templates/startDevWorker/ProxyWorker.ts

```diff
@@ -137,17 +139,46 @@ export class ProxyWorker implements DurableObject {
 					// errors here are network errors or from response post-processing
 					// to catch only network errors, use the 2nd param of the fetch.then()
 
-					void sendMessageToProxyController(this.env, {
-						type: "error",
-						error: {
-							name: error.name,
-							message: error.message,
-							stack: error.stack,
-							cause: error.cause,
-						},
-					});
-
-					deferredResponse.reject(error);
+					// we have crossed an async boundary, so proxyData may have changed
+					const newUserWorkerUrl =
+						this.proxyData && urlFromParts(this.proxyData.userWorkerUrl);
+
+					// only report errors if the downstream proxy has NOT changed
+					if (userWorkerUrl.href === newUserWorkerUrl?.href) {
+						void sendMessageToProxyController(this.env, {
+							type: "error",
+							error: {
+								name: error.name,
+								message: error.message,
+								stack: error.stack,
+								cause: error.cause,
+							},
+						});
+
+						deferredResponse.reject(error);
+					}
+
+					// if the request can be retried (subset of idempotent requests which have no body), requeue it
+					else if (request.method === "GET" || request.method === "HEAD") {
+						this.requestQueue.set(request, deferredResponse);
```

Comment:
Will this not put this request onto the back of the queue after subsequent pending requests?
Don't we want the aborted requests to be retried before any pending requests?

Reply:
I've added 2 commits to do our best to fire off requests by order of arrival but it still comes down to when the .catch is hit – we won't be able to guarantee request order in the case:

1. request1 comes in, request1 proxied
2. reloadStart received, pauses proxying
3. request2 comes in, held in queue
4. reloadComplete received, unpauses proxying, request2 proxied
5. workerd for request1 eventually exits and request1 fails, request1 is requeued

Although I just simulated this scenario delaying miniflare killing the workerd instance by 5 seconds (guaranteeing the new workerd instance starts up earlier). I expected to have to call `this.processQueue` to actually process the request after requeuing it but the test still passed so need to a further look on Monday

Reply:
Is there likely to be a delay before a request errors when the workerd is reloaded? Is the error due to the request timing out (which of course would introduce this delay) or because the connection was closed (which feels like it should error pretty instantly).
In the latter case, it feels like as a new workerd boots up and becomes available to receive messages we should have two lists of requests:
- requests that failed when the connection closed and need retrying
- requests that arrived after the connection closed and are pending
Our logic, therefore could be to put these retry requests in their own list and make sure we empty that when a workerd reload event comes it, before trying to send any of the pending requests.

Of course, if some of these retryable requests do not fail until after the new workerd has loaded and started accepting requests there is not much we can do about that, since the boat has already sailed. But always prioritising retyable requests over pending requests feels like as good as we can offer.

Reply:
Had that last comment drafted and forgot to hit send, noticed later and just hit send but forgot to add that the request order sorting has minimal effect per batch because the requestQueue is processed synchronously and that order would have minimal effect on the order they're received by the UserWorker over the http async boundary. Brendan has also suggested something similar in [another comment](https://github.com/cloudflare/workers-sdk/pull/4867#discussion_r1478184550)

I forgot to hit send originally because I got nerdsniped by the lack of a `processQueue` call. Like I mentioned, I simulated this scenario delaying miniflare killing the workerd instance by 5 seconds. I've since confirmed with @mrbbot that this also delays the starting of the next workerd instance which waits for the previous to terminate. Which confirms why I could never hit the race condition I described above – step 4 is (essentially) guaranteed to happen after step 5.

I confirmed this by timestamping the events and the timing is incredible consistent:
- 13-18ms between `pause` message and `.catch`
- 1060-1070ms between `.catch` and `play` message

Therefore, we don't need to consider the edge case I discussed above.

Further, I pushed those 2 commits to demonstrate (1) the little effort required to order the requests, (2) that the order they are fired still doesn't guarantee the order in which they arrive. I intended to revert the commits because I don't think the added overhead is justified by the only slightly better chance of maintaining the received order. But I like your suggestion of a second queue for requests to retry so I've instead gone with that approach which is just as good (enough) but adds way less overhead.

Reply:
https://github.com/cloudflare/workers-sdk/pull/4867/commits/5288c3c5836ddea6e3b5a84b9e42156ad3c4ec6c